### PR TITLE
[WIP] QNNPACK: Add support for dynamic quantization.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Test Plan: Added new tests to QNNPACK's test suite to cover the new use case.  All new tests are passing.

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x4c2-dq-sse2.c [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    size_t mr,
+    size_t nr,
+    size_t k,
+    const uint8_t* restrict a,
+    size_t a_stride,
+    const void* restrict w,
+    float* restrict c,
+    size_t c_stride,
+    const union pytorch_qnnp_conv_quantization_params
+        quantization_params[RESTRICT_STATIC 1]) {
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x4c2-dq-sse2.c]: We need to pass the multiplier somehow, either inside pytorch_qnnp_conv_quantization_params or as a separate argument.  Let me know if you have a preference.  Other than that, this should be done.

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h [hunk 2]
```diff
@@ -310,6 +310,126 @@ class GemmMicrokernelTester {
...
+
+      const float requantizationScale = 1.0f / float(cScale);
+      const union pytorch_qnnp_conv_quantization_params quantizationParams =
+          pytorch_qnnp_compute_conv_quantization_params(
+              aZeroPoint(),
+              bZeroPoint(),
+              requantizationScale,
+              cZeroPoint,
+              qmin(),
+              qmax());
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h]: @supriyar I still need to calculate pytorch_qnnp_conv_quantization_params even though only the input and kernel zero points are getting used.  Please let me know what API works best for you.  Maybe it is just enough to pass in the input and kernel zero points and the multiplier as three arguments.  Let me know what you prefer.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x4c2-dq-sse2.c [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <immintrin.h>
+
+#include <qnnpack/q8gemm.h>
+#include <requantization/runtime-sse2.h>
+
+void pytorch_q8gemm_dq_ukernel_4x4c2__sse2(
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x4c2-dq-sse2.c]: Note: Dynamic quantization kernels have a __dq__ in their name.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit bd2974e0:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=bd2974e093cbd7f1d67ecb0cd939d95c720db67d).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4144219) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80878473) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80878473))
<details>
<summary>
<code>Jan 06 21:48:18 CMake Error at cmake/Dependencies.cmake:482 (add_subdirectory):</code>
</summary>

```
Jan 06 21:48:18 -- Brace yourself, we are building NNPACK 
Jan 06 21:48:18 -- Performing Test NNPACK_ARCH_IS_X86_32 
Jan 06 21:48:18 -- Performing Test NNPACK_ARCH_IS_X86_32 - Failed 
Jan 06 21:48:18 -- NNPACK backend is x86-64 
Jan 06 21:48:18 -- Caffe2: Cannot find gflags automatically. Using legacy find. 
Jan 06 21:48:18 -- Found gflags: /usr/include   
Jan 06 21:48:18 -- Caffe2: Found gflags  (include: /usr/include, library: /usr/lib/x86_64-linux-gnu/libgflags.so) 
Jan 06 21:48:18 -- Caffe2: Cannot find glog automatically. Using legacy find. 
Jan 06 21:48:18 -- Found glog: /usr/include   
Jan 06 21:48:18 -- Caffe2: Found glog (include: /usr/include, library: /usr/lib/x86_64-linux-gnu/libglog.so) 
Jan 06 21:48:18 CMake Error at cmake/Dependencies.cmake:482 (add_subdirectory): 
Jan 06 21:48:18   The binary directory 
Jan 06 21:48:18  
Jan 06 21:48:18     /var/lib/jenkins/workspace/build/third_party/googletest 
Jan 06 21:48:18  
Jan 06 21:48:18   is already used to build a source directory.  It cannot be used to build 
Jan 06 21:48:18   source directory 
Jan 06 21:48:18  
Jan 06 21:48:18     /var/lib/jenkins/workspace/third_party/googletest 
Jan 06 21:48:18  
Jan 06 21:48:18   Specify a unique binary directory name. 
```

</details>

#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4144197) pytorch_linux_xenial_py2_7_9_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80878534) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80878534))
<details>
<summary>
<code>Jan 06 21:48:17 CMake Error at cmake/Dependencies.cmake:482 (add_subdirectory):</code>
</summary>

```
Jan 06 21:48:16 -- MKL OpenMP library: -fopenmp 
Jan 06 21:48:16 -- The ASM compiler identification is GNU 
Jan 06 21:48:16 -- Found assembler: /opt/cache/bin/cc 
Jan 06 21:48:17 -- Check if compiler accepts -pthread 
Jan 06 21:48:17 -- Check if compiler accepts -pthread - yes 
Jan 06 21:48:17 -- Found PythonInterp: /opt/python/2.7.9/bin/python (found version "2.7.9")  
Jan 06 21:48:17 -- Brace yourself, we are building NNPACK 
Jan 06 21:48:17 -- Performing Test NNPACK_ARCH_IS_X86_32 
Jan 06 21:48:17 -- Performing Test NNPACK_ARCH_IS_X86_32 - Failed 
Jan 06 21:48:17 -- NNPACK backend is x86-64 
Jan 06 21:48:17 CMake Error at cmake/Dependencies.cmake:482 (add_subdirectory): 
Jan 06 21:48:17   The binary directory 
Jan 06 21:48:17  
Jan 06 21:48:17     /var/lib/jenkins/workspace/build/third_party/googletest 
Jan 06 21:48:17  
Jan 06 21:48:17   is already used to build a source directory.  It cannot be used to build 
Jan 06 21:48:17   source directory 
Jan 06 21:48:17  
Jan 06 21:48:17     /var/lib/jenkins/workspace/third_party/googletest 
Jan 06 21:48:17  
Jan 06 21:48:17   Specify a unique binary directory name. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h (hunk 2 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h]: Yes, I think having the pytorch_qnnp_conv_quantization_params is an overkill. We can just pass in the zero_points and the multiplier as arguments to the kernel directly.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x4c2-dq-sse2.c (hunk 1 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x4c2-dq-sse2.c]: I'd vote for separate argument. See comment below.

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h [hunk 4]
```diff
@@ -310,6 +319,106 @@ class GemmMicrokernelTester {
...
+          for (size_t kIndex = 0; kIndex < k(); kIndex++) {
+            ASSERT_LE(n(), packedN());
+            ASSERT_LT(mIndex * n() + nIndex, acc.size());
+            ASSERT_LT(mIndex * k() + kIndex, a.size());
+            acc[mIndex * n() + nIndex] +=
+                (int32_t(aPtr[mIndex * aStride() + kIndex]) -
+                 int32_t(aZeroPoint())) *
+                (int32_t(b[nIndex * k() + kIndex]) - int32_t(bZeroPoint()));
+          }
+          acc[mIndex * n() + nIndex] = acc[mIndex * n() + nIndex] * multiplier() + bias[nIndex];
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h]: @supriyar This is the reference impl.  Please check the math matches what dynamic quantization expects.

File: aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h [hunk 5]
```diff
@@ -310,6 +319,106 @@ class GemmMicrokernelTester {
...
+        }
+      }
+
+      const struct pytorch_qnnp_conv_dynamic_quantization_params quantizationParams{
+        aZeroPoint(),
+        bZeroPoint(),
+        multiplier(),
+      };
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h]: @supriyar Example usage.

**Comment by kimishpatel** [comment]
@AshkanAliabadi, @supriyar, my general feeling is that if we can reuse the existing kernel that would be great. So for example if we move all the bias add part, outside of the main kernel then dynamic quantization and static can use the main subkernel. The perf hit due to this should not be huge I imagine.

**Comment by AshkanAliabadi** [comment]
I agree that it would be preferable to avoid code duplication, but consolidating these functions is not as straightforward as it seems especially in a way that doesn't hurt (a) performance, and (b) code readability.  There are minor differences between the kernels such as what registers to save or reuse, etc., and one needs to either turn to compile-time macro magic, or run-time conditions to unify these code paths, and I am not sure either of those is a clear win.  I would like to move forward with this implementation in the favor of time and we can always revisit to improve things if need be.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S [hunk 6]
```diff
@@ -11,7 +11,7 @@
 
 .syntax unified
 
-# void pytorch_q8gemm_ukernel_4x8__aarch32_neon(
+# void pytorch_q8gemm_dq_ukernel_4x8__aarch32_neon(
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S]: Does this file need to change?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/params.h [hunk 7]
```diff
@@ -274,6 +280,18 @@ typedef void (*pytorch_q8gemm_ukernel_function)(
     size_t c_stride,
     const union pytorch_qnnp_conv_quantization_params* quantization_params);
 
+typedef void (*pytorch_q8gemm_dq_ukernel_function)(
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/params.h]: Can you add a comment here briefly describing what this kernel does

**Comment by supriyar** [comment]
LGTM mostly, can you describe the test plan in the PR description?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/params.h (hunk 7 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/qnnpack/params.h]: Thanks, done.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S (hunk 6 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/q8gemm/4x8-aarch32-neon.S]: Good catch, removed.

**Pull request closed without merging**